go/adbc/driver/bigquery/record_reader.go

```diff
@@ -0,0 +1,315 @@
+	}
+	return ctx.Err()
+}
+
+func runQuery(ctx context.Context, query *bigquery.Query, executeUpdate bool) (bigquery.ArrowIterator, int64, error) {
+	job, err := query.Run(ctx)
+	if err != nil {
+		return nil, -1, err
+	}
+	if executeUpdate {
+		return nil, 0, nil
+	}
+
+	iter, err := job.Read(ctx)
+	if err != nil {
+		return nil, -1, err
+	}
+	arrowIterator, err := iter.ArrowIterator()
+	if err != nil {
+		return nil, -1, err
+	}
+	totalRows := int64(iter.TotalRows)
+	return arrowIterator, totalRows, nil
+}
+
+func ipcReaderFromArrowIterator(arrowIterator bigquery.ArrowIterator, alloc memory.Allocator) (*ipc.Reader, error) {
+	arrowItReader := bigquery.NewArrowIteratorReader(arrowIterator)
+	return ipc.NewReader(arrowItReader, ipc.WithAllocator(alloc))
+}
+
+func getQueryParameter(values arrow.Record, row int, parameterMode string) ([]bigquery.QueryParameter, error) {
+	parameters := make([]bigquery.QueryParameter, values.NumCols())
+	includeName := parameterMode == OptionValueQueryParameterModeNamed
+	schema := values.Schema()
+	for i, v := range values.Columns() {
+		pi, err := arrowValueToQueryParameterValue(schema.Field(i), v, row)
+		if err != nil {
+			return nil, err
+		}
+		parameters[i] = pi
+		if includeName {
+			parameters[i].Name = values.ColumnName(i)
+		}
+	}
+	return parameters, nil
+}
+
+func runPlainQuery(ctx context.Context, query *bigquery.Query, alloc memory.Allocator, resultRecordBufferSize int) (bigqueryRdr *reader, totalRows int64, err error) {
+	arrowIterator, totalRows, err := runQuery(ctx, query, false)
+	if err != nil {
+		return nil, -1, err
+	}
+	rdr, err := ipcReaderFromArrowIterator(arrowIterator, alloc)
+	if err != nil {
+		return nil, -1, err
+	}
+
+	chs := make([]chan arrow.Record, 1)
+	ctx, cancelFn := context.WithCancel(ctx)
+	ch := make(chan arrow.Record, resultRecordBufferSize)
+	chs[0] = ch
+
+	defer func() {
+		if err != nil {
+			close(ch)
+			cancelFn()
+		}
+	}()
+
+	bigqueryRdr = &reader{
+		refCount:   1,
+		chs:        chs,
+		curChIndex: 0,
+		err:        nil,
+		cancelFn:   cancelFn,
+		schema:     nil,
+	}
+
+	go func() {
+		defer rdr.Release()
+		for rdr.Next() && ctx.Err() == nil {
+			rec := rdr.Record()
+			rec.Retain()
+			ch <- rec
+		}
+
+		err = checkContext(ctx, rdr.Err())
+		defer close(ch)
+	}()
+	return bigqueryRdr, totalRows, nil
+}
+
+// kicks off a goroutine for each endpoint and returns a reader which
+// gathers all of the records as they come in.
+func newRecordReader(ctx context.Context, query *bigquery.Query, boundParameters array.RecordReader, parameterMode string, alloc memory.Allocator, resultRecordBufferSize, prefetchConcurrency int) (bigqueryRdr *reader, totalRows int64, err error) {
+	if boundParameters == nil {
+		return runPlainQuery(ctx, query, alloc, resultRecordBufferSize)
+	}
+
+	recs := make([]arrow.Record, 0)
+	for boundParameters.Next() {
+		rec := boundParameters.Record()
+		recs = append(recs, rec)
+	}
```

Comment:
why iterate the bound parameter stream and store all the records ahead of time instead of iterating the stream concurrently with performing the queries?

Also, if we're going to do it this way, you need to call `rec.Retain()` before putting it into the slice, and then call `Release` on the recs as you iterate over them. Each call to `Record()` on the record reader is allowed to release the previous record, so you need to call `Retain()` so that it survives in the slice

Reply:
Hi @zeroshade, thanks for these advices! I was looking at snowflake's implementation as reference, and here if I'm not wrong, the idea is to count how many channels do we need so that each Record batch and their results can be put in a separate channel (and the requests would be in their own goroutine).

```go
batches := int64(len(recs))
chs := make([]chan arrow.Record, batches)
```

Reply:
I hope this makes sense but please do let me know if we should do this with another approach :)

Reply:
I think the strategy for parallelism here is a little different from the Snowflake implementation you're comparing it to. In Snowflake, a single query results in an `ArrowStreamLoader` which contains potentially multiple streams that are all known up front. These multiple streams are what the recordReader uses for its channels. It can _also_ handle multiple batches of query parameters but does not do so concurrently. For these it iterates through them one by one, executing a query for each (see snowflake's `concatReader`).

BigQuery can expose result sets as multiple streams when using certain APIs, but it's not the case here. The current `ArrowIterator` interface only exposes a single stream so it's probably best to just keep things sequential for now. So I would say that for each `boundParameters` batch, lets run a single query that returns records without any parallelism.

Reply:
Ohh okay I got i! I'll change the code soon!